### ISSUE TITLE
check scaling of docker

### ISSUE THREAD

**Issue by kdaily** [Sage-Bionetworks/Agora#289]
@Zaknarfen will look at what can be done with docker for performance.

**Comment by blogsdon** [comment]
@Zaknarfen what's the status on this?

**Comment by Zaknarfen** [comment]
I've spent quite some time looking into this by now and here is what I gathered. 

- The documentation says we can use a `.json` file to configure things the way we want to. This is mostly applied to `Docker` usage itself (we are using `Docker-compose`).

- For `Docker-compose` we can simply add a deploy property to the services we want, e.g.:

```
version: '3'
services:
  redis:
    image: redis:alpine
    deploy:
      resources:
        limits:
          cpus: '0.50'
          memory: 50M
        reservations:
          cpus: '0.25'
          memory: 20M
```

The problem is that we don't run the latest `Docker-compose` version. We run version `2` and it is **fairly** complicated to move to version `3` (I tried at first and couldn't do it). In version `2` there is no `deploy` property.

- Furthermore, `Docker` memory usage seems to be different based on the OS. For our luck, according to [this](https://stackoverflow.com/questions/44533319/how-to-assign-more-memory-to-docker-container) question, `Linux` machines are used to the fullest by `Docker` so we should be ok even without `deploy`.

- A snapshot of what we are using at the moment:

![image](https://user-images.githubusercontent.com/4223973/43660597-dd91d928-9735-11e8-8739-bbd7fefed1b5.png)

I believe we are really good for a high volume of traffic with this at the moment, and we need to consider two things:

1. Move our stack to AWS Elastic Beanstalk (probably can be done in different ways). That would mean changing from one EC2 instance with multiple containers to the same containers scaled in an Amazon ECS cluster in an Elastic Beanstalk environment. AWS would do the balancing for us (no scaling problems here). We would just need to see how that plays with `Nginx`, do we keep four applications per instance?

2. Keep our version or change to the latest `Docker-compose` version. This would mean a complete change in how we do everything in `Docker`, like networks, connections, volumes, deprecated stuff we rely on etc... I don't know if we need this just to add `deploy` and use `Docker` in `swarm` mode. When you run `Docker` in `swarm` mode you are trying to create, manage and automate the cluster itself (one step above compose, that only orchestrates containers in one stack), but... can we achieve the same thing with compose in AWS Elastic Beanstalk? probably. I am not convinced we need to update, here is a more [complete](https://www.quora.com/Why-do-we-use-docker-swarm) answer.
